docs/RFCS/20201020_sql_query_denylist.md

```diff
@@ -0,0 +1,183 @@
+
+The administrator may modify the file content as needed; and if the entire denylist is no longer available, they may remove it entirely. The changes will take effect immediately.
+
+### User experience from a developer's point of view
+A developer may be sending a query that has been denied on the denylist or would be denied and is currently on the staging denylist. In the latter case, they will also be able to view the warning on console that this query is still executing, but would be denied due to a pattern match on a staging denylist. In the former case, they will receive an error message that "query matched a pattern in the denylist by the database administrator" with the specific regex query pattern that it matched. 
+
+# Reference-level explanation
+
+## Detailed design
+
+### Design overview
+We will implement a file-managed denylist, stored locally on each machine as a YAML configuration file. The exact location of the denylist should be configured with a flag – this will allow users to specify the location of their file, such as via the k8s ConfigMap for CockroachCloud SREs. The file location will be watched so that as soon as the config file is created, modified, or removed, the denylist will be updated. Pattern matching between queries executed and queries on the denylist will be done via regex so users maintain granular control of the queries affected by the denylist.
+
+A staging YAML file will be supported so users will be able to confirm the queries that would be denied: when the staging file is set, users will see logs of the queries that are not yet denied but will be denied if the non-staging denylist file is set.  
+
+### Configuring denylist location
+A denylist location flag can be configured in [pkg/cli/flags.go](https://github.com/cockroachdb/cockroach/blob/master/pkg/cli/flags.go). eg. `--denylist=’denylist.yaml’`. If this flag has not been set, then there will be no location to watch for a denylist and therefore the functionality is off by default. 
+
+### File-managed denylist implementation
+We expect users to create their own YAML file with a `sql` field that lists an array of strings representing regex to match any SQL queries to be denied. This allows granular control of specific command types as well as database objects — a user should be able to denylist any command of a certain type, any command pertaining to a particular database object, or any additional parameters that may be affecting the cluster.
+
+In the future, we could extend the functionality with additional fields such as, for example, `IP` to support denying queries made from specific IP addresses. 
+
+Example denylist: **denylist.yaml**
+```
+# Queries to be denied; in the future we may support denying other fields.
```

Comment:
What happens if a user enters an invalid entry?

Reply:
That's fair. I'd suggest that an invalid entry (specifically an invalid key, one that's not `sql` in this case) be surfaced as a warning/error to console and the logs so that the administrator knows that the denylist was not processed correctly. @otan thoughts here?